File: 2015/03_Perfectly_Spherical_Houses_in_a_Vacuum/part_2/main.js
/**
 * Author: happ_2h https://github.com/happ-2h
 *
 * Solution for Advent of Code 2015 Day 3.2
 */

let input = null;

try {
  input = require('fs').readFileSync("../input.txt", "utf-8").split("");
} catch(err) {
  console.error(err);
}

if (input !== null) {
  // Coordinates
  const santa = { x: 0, y: 0 };
  const robot = { x: 0, y: 0 };

  const visited = new Set();
  // Save starting point
  visited.add(JSON.stringify(santa));
  visited.add(JSON.stringify(robot));

  for (let i = 0; i < input.length; ++i) {
    switch(input[i]) {
      // Even: Santa moves, odd: robot moves
      case "^":
        i&1 ? --robot.y : --santa.y;
        break;
      case "v":
        i&1 ? ++robot.y : ++santa.y;
        break;
      case "<":
        i&1 ? --robot.x : --santa.x;
        break;
      case ">":
        i&1 ? ++robot.x : ++santa.x;
        break;
    }
    i&1 ? visited.add(JSON.stringify(robot)) : visited.add(JSON.stringify(santa));
  }

  console.log(visited.size);
}
else console.error("Failed to read input.txt");

File: 2015/03_Perfectly_Spherical_Houses_in_a_Vacuum/part_2/main.kt
/**
 * Author: happ_2h https://github.com/happ-2h
 *
 * Solution for Advent of Code 2015 Day 3.2
 */

import java.io.File

data class Vec2d(var x: Int, var y: Int)

fun main(args: Array<String>) {
  val path = File("../input.txt").readText()

  // Keep track of Cartesian coordinates
  val santa: Vec2d = Vec2d(0, 0)
  val robot: Vec2d = Vec2d(0, 0)

  // Save starting point
  val visited: MutableSet<String> = mutableSetOf(santa.toString())

  // Move based on input
  for (i in 0 until path.length) {
    // Even: Santa moves. Odd: robot moves
    when (path[i]) {
      '^' -> if (i and 1 == 0) --robot.y else --santa.y
      'v' -> if (i and 1 == 0) ++robot.y else ++santa.y
      '<' -> if (i and 1 == 0) --robot.x else --santa.x
      '>' -> if (i and 1 == 0) ++robot.x else ++santa.x
    }

    // Save result (unique)
    visited.add(if (i and 1 == 0) robot.toString() else santa.toString())
  }

  println(visited.size)
}